File: cpp/api/src/base/PingableSet.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include "PingableSet.h"
#include "PingableObject.h"
#include "Logger.h"
#include <iostream>

namespace cameo {

PingableSet::PingableSet() {
}

void PingableSet::add(PingableObject * object) {

	std::lock_guard<std::mutex> lock {m_mutex};

	m_set.insert(object);

	Logger::log(std::string{"Inserted"} + object->toString());
}

void PingableSet::remove(PingableObject * object) {

	std::lock_guard<std::mutex> lock {m_mutex};

	Logger::log(std::string{"Removing "} + object->toString());

	std::set<PingableObject *>::iterator it = m_set.find(object);

	if (it != m_set.end()) {
		m_set.erase(it);

		Logger::log(std::string{"Removed "} + object->toString());
	}
}

void PingableSet::pingAll(int timeout) {

	Logger::log(std::string{"Ping all "} + std::to_string(timeout));

	std::lock_guard<std::mutex> lock {m_mutex};

	for (std::set<PingableObject *>::iterator it = m_set.begin(); it != m_set.end(); ++it) {
		if ((*it)->isPinged()) {
			Logger::log(std::string{"Ping "} + (*it)->toString());
			bool pong = (*it)->ping(timeout);
			if (pong) {
				Logger::log(std::string{"Pong for "} + (*it)->toString());
			}
			else {
				Logger::log(std::string{"No pong for "} + (*it)->toString());
			}
		}
	}
}

}

File: java/api/src/main/java/eu/ill/cameo/api/base/PingableSet.java
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

package eu.ill.cameo.api.base;


import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

/**
 * Class defining a set of IPingables.
 */
public class PingableSet {

	private Set<IPingable> pingableSet = new HashSet<IPingable>();

	/**
	 * Adds a IPingable object.
	 * @param pingable The IPingable object.
	 */
	public void add(IPingable pingable) {
		synchronized (pingableSet) {
			pingableSet.add(pingable);
		}
		
		System.out.println("Added pingable " + pingable);
	}
	
	/**
	 * Removes a IPingable object.
	 * @param pingable The IPingable object.
	 */
	public void remove(IPingable pingable) {
		synchronized (pingableSet) {
			pingableSet.remove(pingable);
		}	
		
		System.out.println("Removed pingable " + pingable);
	}
	
	/**
	 * Pings all the IPingable objects.
	 * @param timeout The timeout.
	 */
	public void pingAll(int timeout) {
		
		synchronized (pingableSet) {
			// The iteration must be in the synchronized block.
			Iterator<IPingable> i = pingableSet.iterator();
			while (i.hasNext()) {
				IPingable pingable = i.next();
				System.out.println("Pinging " + pingable);
				boolean pong = pingable.ping(timeout);
				
				if (pong) {
					System.out.println("Pong for " + pingable);
				}
				else {
					System.out.println("No pong for " + pingable);
				}
			}
		}
	}
	
}